lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/auth/OAuthRefreshTokenDaoImpl.java

```diff
@@ -85,6 +86,21 @@
 			+ " WHERE " + COL_OAUTH_REFRESH_TOKEN_PRINCIPAL_ID + " = :" + PARAM_PRINCIPAL_ID
 			+ " AND " + COL_OAUTH_REFRESH_TOKEN_CLIENT_ID + " = :" + PARAM_CLIENT_ID;
 
+	private static final String SELECT_ACTIVE_TOKEN_COUNT = "SELECT COUNT(*) FROM " + TABLE_OAUTH_REFRESH_TOKEN
+			+ " WHERE " + COL_OAUTH_REFRESH_TOKEN_PRINCIPAL_ID + " = :" + PARAM_PRINCIPAL_ID
+			+ " AND " + COL_OAUTH_REFRESH_TOKEN_CLIENT_ID + " = :" + PARAM_CLIENT_ID
+			+ " AND " + COL_OAUTH_REFRESH_TOKEN_LAST_USED + " > (NOW() - INTERVAL  :" + PARAM_DAYS + " DAY)";
+
+	private static final String DELETE_LEAST_RECENTLY_USED_ACTIVE_TOKEN = 
+			"DELETE FROM " + TABLE_OAUTH_REFRESH_TOKEN
+				+ " WHERE " + COL_OAUTH_REFRESH_TOKEN_PRINCIPAL_ID + " = :" + PARAM_PRINCIPAL_ID
+				+ " AND " + COL_OAUTH_REFRESH_TOKEN_CLIENT_ID + " = :" + PARAM_CLIENT_ID
+				+ " AND " + COL_OAUTH_REFRESH_TOKEN_LAST_USED
+				+ " > (NOW() - INTERVAL  :" + PARAM_DAYS + " DAY)"
+				+ " ORDER BY " + COL_OAUTH_REFRESH_TOKEN_LAST_USED + " ASC"
```

Comment:
instead of
" AND lastUsed > threshold ORDER BY  lastUsed ASC LIMIT 1"
what do you think of
"ORDER BY lastUsed DESC OFFSET 100"
(where 100 is just an example of the number of tokens you want to retain).
I.e. this deletes all but the most recent 100 tokens.  It has three advantages:
(1) You don't need to know the count before executing the query;
(2) There is no 'race condition' in which you check the count and then it changes (a token expires) just before you run the delete query;
(3) It cleans up old, expired tokens.
Essentially the operation is idempotent.  It just cleans up all but the most recent 100 tokens.

Reply:
I think this is a great idea. I will make that change.

Reply:
I think that the COUNT query introduced in this PR is also obsoleted by this change (there is no need to get a count if we have a "safe" DELETE query), so I will remove that as well.